neonvm/runner/main.go

```diff
@@ -807,6 +815,86 @@ func listenForCPUChanges(ctx context.Context, logger *zap.Logger, port int32, cg
 	}
 }
 
+func drainLogsReader(reader *bufio.Reader) error {
+	for {
+		slice, err := reader.ReadSlice('\n')
+		if err != nil {
+			if len(slice) != 0 {
+				// Print and add a newline to the slice we got
+				fmt.Printf("%s\n", slice)
+			}
+
+			return err
+		}
+
+		fmt.Printf("%s", slice)
+	}
+}
+
+// writes from socket to stdout line by line
+func forwardLogs(ctx context.Context, logger *zap.Logger, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	delay := 3 * time.Second
+	var conn net.Conn
+	var reader *bufio.Reader
+	b := &backoff.Backoff{
+		Min:    100 * time.Millisecond,
+		Max:    delay,
+		Factor: 2,
+		Jitter: true,
+	}
+
+	tryForward := func() {
+		if conn == nil {
+			var err error
+			conn, err = net.Dial("unix", logSerialSocket)
+			if err != nil {
+				logger.Error("failed to dial to logSerialSocket", zap.Error(err))
+				return
+			}
+			reader = bufio.NewReader(conn)
+		}
+
+		b.Attempt()
+		err := conn.SetReadDeadline(time.Now().Add(delay))
+		if err != nil {
+			logger.Error("failed to set read deadline", zap.Error(err))
+			conn = nil
+			return
+		}
+
+		err = drainLogsReader(reader)
+		if errors.Is(err, os.ErrDeadlineExceeded) {
+			// We've hit the deadline, meaning the reading session was successful.
+			b.Reset()
+			return
+		}
+
+		if err != nil {
+			logger.Error("failed to read from socket", zap.Error(err))
+			conn = nil
+		}
```

Comment:
Should we handle (EOF if ctx.Err() != nil) here and below? When trying this locally, the final logs are:

```
{"level":"info","ts":1706032966.0483866,"logger":"neonvm-runner.terminate-qemu-on-sigterm","caller":"runner/main.go:1111","msg":"got signal, sending powerdown command to QEMU"}
{"level":"info","ts":1706032966.0497675,"logger":"neonvm-runner.terminate-qemu-on-sigterm","caller":"runner/main.go:1132","msg":"system_powerdown command sent to QEMU"}
{"timestamp":"2024-01-23T18:02:46.051992Z","level":"ERROR","fields":{"message":"cgroup watcher terminated"},"target":"vm_monitor::runner","filename":"libs/vm_monitor/src/runner.rs","line_number":206}
The system is going down NOW!
Sent SIGTERM to all processes
udhcpc: received SIGTERM
2024-01-23T18:02:46.966903Z  INFO vector: Vector has stopped.
exiting
Sent SIGKILL to all processes
Requesting system poweroff
[   11.666647] ACPI: PM: Preparing to enter system sleep state S5
[   11.667749] reboot: Power down
{"level":"error","ts":1706032969.183614,"logger":"neonvm-runner","caller":"runner/main.go:875","msg":"failed to read from socket","error":"EOF","stacktrace":"main.forwardLogs.func1\n\t/workspace/neonvm/runner/main.go:875\nmain.forwardLogs\n\t/workspace/neonvm/runner/main.go:881"}
{"level":"info","ts":1706032969.2671506,"logger":"neonvm-runner","caller":"runner/main.go:718","msg":"QEMU exited without error"}
{"level":"error","ts":1706032969.2673178,"logger":"neonvm-runner","caller":"runner/main.go:890","msg":"failed to read from socket","error":"EOF","stacktrace":"main.forwardLogs\n\t/workspace/neonvm/runner/main.go:890"}
{"level":"info","ts":1706032969.2674258,"logger":"neonvm-runner","caller":"runner/main.go:814","msg":"shut down cpu_change server"}
```

note the "failed to read from socket" error logs, with "EOF" as the error.